File: app_codebase/dart_bkend_base_02/bin/core/http/response/response.dart
/// Shared JSON format for every API response: either `ok` plus `data`, or `ok` false plus an
/// `error` object with `code` and `message`.
///
/// The functions [jsonOk] and [jsonError] build full Shelf [Response] values with the right
/// content type. Route handlers use them directly or through [ShelfJsonHttpResponses]; middleware
/// and the “no such route” path use them too so clients always see the same structure.
///
/// [httpResponses] is a single object you can pass anywhere an [HttpResponseContract] is
/// expected.
library;

import 'dart:convert';

import 'package:shelf/shelf.dart';

import '../contracts/http_response_contract.dart';

Map<String, Object?> jsonSuccessBody(Object? data) => {
      'ok': true,
      'data': data,
    };

Map<String, Object?> jsonErrorBody({
  required String code,
  required String message,
}) =>
    {
      'ok': false,
      'error': {
        'code': code,
        'message': message,
      },
    };

Response _shelfJsonOk(Object? data, {int status = 200}) => Response(
      status,
      body: jsonEncode(jsonSuccessBody(data)),
      headers: {'content-type': 'application/json; charset=utf-8'},
    );

Response _shelfJsonError({
  required String code,
  required String message,
  int status = 400,
}) =>
    Response(
      status,
      body: jsonEncode(jsonErrorBody(code: code, message: message)),
      headers: {'content-type': 'application/json; charset=utf-8'},
    );

/// JSON success reply with optional HTTP status. Core code and guards call this; feature modules
/// usually go through [HttpResponseContract] instead.
Response jsonOk(Object? data, {int status = 200}) =>
    _shelfJsonOk(data, status: status);

/// JSON error reply: you choose the HTTP status; the body uses the same overall JSON shape as
/// success, with `ok: false` and an `error` block.
Response jsonError({
  required String code,
  required String message,
  int status = 400,
}) =>
    _shelfJsonError(code: code, message: message, status: status);

/// Delegates to [jsonOk] / [jsonError] so [HttpResponseContract] can be satisfied without extra
/// duplication.
final class ShelfJsonHttpResponses implements HttpResponseContract {
  const ShelfJsonHttpResponses();

  @override
  Response jsonOk(Object? data, {int status = 200}) =>
      _shelfJsonOk(data, status: status);

  @override
  Response jsonError({
    required String code,
    required String message,
    int status = 400,
  }) =>
      _shelfJsonError(code: code, message: message, status: status);
}

/// Default [HttpResponseContract] instance for passing into feature registration functions.
const HttpResponseContract httpResponses = ShelfJsonHttpResponses();

File: app_codebase/python_base_05/bin/core/http/response/response.py
"""Shared JSON format for every API response: either ``ok`` plus ``data``, or ``ok`` false plus an
``error`` object with ``code`` and ``message``.

The free functions ``json_ok`` and ``json_error`` build full ``JSONResponse`` objects with the
right content type. Route handlers use them directly or through :class:`JsonHttpResponses`; the
middleware and the “no such route” path use them too so clients always see the same structure.

``http_responses`` is a single object you can pass anywhere an :class:`~core.http.contracts.response_contract.HttpResponseContract` is expected.
"""

from typing import Any, Mapping

from starlette.responses import JSONResponse, Response

from core.http.contracts.response_contract import HttpResponseContract


def json_success_body(data: object | None) -> dict[str, Any]:
    return {"ok": True, "data": data}


def json_error_body(*, code: str, message: str) -> dict[str, Any]:
    return {"ok": False, "error": {"code": code, "message": message}}


def json_ok(data: object | None, *, status: int = 200) -> Response:
    body: Mapping[str, Any] = json_success_body(data)
    return JSONResponse(content=dict(body), status_code=status)


def json_error(*, code: str, message: str, status: int = 400) -> Response:
    body = json_error_body(code=code, message=message)
    return JSONResponse(content=body, status_code=status)


class JsonHttpResponses:
    """Thin wrapper so ``HttpResponseContract`` can be satisfied by delegating to ``json_ok`` /
    ``json_error`` below.
    """

    def json_ok(self, data: object | None, *, status: int = 200) -> Response:
        return json_ok(data, status=status)

    def json_error(self, *, code: str, message: str, status: int = 400) -> Response:
        return json_error(code=code, message=message, status=status)


http_responses: HttpResponseContract = JsonHttpResponses()
